File: sdk/python/memg/consolidator.py
"""Consolidation: cluster old event facts into pattern facts."""

from __future__ import annotations

import logging
from collections import defaultdict
from datetime import datetime, timedelta, timezone
from typing import Callable, List

from .extract import content_key
from .store import Store
from .types import Fact, FactFilter

logger = logging.getLogger("memg")


def consolidate_entity(
    store: Store,
    embedder,
    llm_chat: Callable[[str], str],
    entity_uuid: str,
) -> int:
    """Consolidate old event facts into pattern facts for a single entity.

    Logic matches Go memory/consolidator.go:
    1. Query event facts older than 30 days, current status, significance <= 5.
    2. Group by tag.
    3. For groups with 3+ facts, ask LLM to produce a behavioral pattern.
    4. Embed the pattern, insert as pattern fact with type="pattern", significance=5.
    5. Mark originals as historical.
    6. Return count of patterns created.
    """
    cutoff = (datetime.now(timezone.utc) - timedelta(days=30)).strftime("%Y-%m-%dT%H:%M:%S")

    filt = FactFilter(
        types=["event"],
        statuses=["current"],
        exclude_expired=True,
        reference_time_before=cutoff,
        max_significance=5,
    )

    facts = store.list_facts_filtered(entity_uuid, filt, limit=500)
    if len(facts) < 3:
        return 0

    by_tag: dict = defaultdict(list)
    for f in facts:
        tag = f.tag if f.tag else "_untagged"
        by_tag[tag].append(f)

    patterns_created = 0

    for tag, group in by_tag.items():
        if len(group) < 3:
            continue

        group.sort(key=lambda f: f.content)

        lines = []
        for f in group:
            lines.append(f"- {f.content}")
        events_text = "\n".join(lines)

        prompt = (
            f"Summarize these {len(group)} related events into a single behavioral pattern statement.\n"
            "The pattern should describe a recurring behavior or tendency.\n"
            "Return ONLY the pattern statement, nothing else.\n"
            "If these events don't form a meaningful pattern, respond with exactly: NONE\n\n"
            f"Events:\n{events_text}"
        )

        try:
            pattern = llm_chat(prompt)
        except Exception as e:
            logger.warning("memg consolidator: llm call for entity %s tag %r: %s", entity_uuid, tag, e)
            continue

        pattern = pattern.strip()
        if not pattern or pattern.upper() == "NONE":
            continue

        try:
            vectors = embedder.embed([pattern])
        except Exception as e:
            logger.warning("memg consolidator: embed pattern for entity %s tag %r: %s", entity_uuid, tag, e)
            continue

        if not vectors:
            continue

        pattern_fact = Fact(
            uuid="",
            content=pattern,
            embedding=vectors[0],
            fact_type="pattern",
            temporal_status="current",
            significance=5,
            tag=tag,
            content_key=content_key(pattern),
            embedding_model=embedder.model_name() if hasattr(embedder, "model_name") else "",
        )

        store.insert_fact(entity_uuid, pattern_fact)

        for f in group:
            store.update_temporal_status(f.uuid, "historical")

        logger.info(
            "memg consolidator: created pattern from %d events (entity=%s tag=%s): %s",
            len(group), entity_uuid, tag, pattern,
        )
        patterns_created += 1

    return patterns_created

File: sdk/typescript/src/consolidator.ts
/**
 * Consolidator: clusters old, low-significance event facts into pattern facts.
 * Matches the Go memory/consolidator.go logic.
 */

import { randomUUID } from 'crypto';
import type { Embedder } from './embedder.js';
import { defaultContentKey, type Store } from './store.js';
import type { Fact, FactFilter } from './types.js';

export interface ConsolidatorConfig {
  store: Store;
  embedder: Embedder;
  llmChat: (prompt: string) => Promise<string>;
  intervalMs?: number;
}

/**
 * Consolidate old event facts for a single entity into pattern facts.
 *
 * Logic (matches Go consolidateEntity):
 * 1. Query event facts older than 30 days, current status, significance <= 5
 * 2. Group by tag
 * 3. For groups with 3+ facts, ask LLM to produce a behavioral pattern
 * 4. Embed the pattern, insert as pattern fact
 * 5. Mark originals as historical
 * 6. Return count of patterns created
 */
export async function consolidateEntity(
  store: Store,
  embedder: Embedder,
  llmChat: (prompt: string) => Promise<string>,
  entityUuid: string
): Promise<number> {
  const cutoff = new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString();

  const filter: FactFilter = {
    types: ['event'],
    statuses: ['current'],
    excludeExpired: true,
    referenceTimeBefore: cutoff,
    maxSignificance: 5,
  };

  const facts = await store.listFactsFiltered(entityUuid, filter, 500);
  if (facts.length < 3) return 0;

  const byTag = new Map<string, Fact[]>();
  for (const f of facts) {
    const tag = f.tag || '_untagged';
    const group = byTag.get(tag) ?? [];
    group.push(f);
    byTag.set(tag, group);
  }

  let patternsCreated = 0;

  for (const [tag, group] of byTag) {
    if (group.length < 3) continue;

    group.sort((a, b) => a.content.localeCompare(b.content));

    const contents = group.map((f) => `- ${f.content}`).join('\n');

    const prompt = `Summarize these ${group.length} related events into a single behavioral pattern statement.
The pattern should describe a recurring behavior or tendency.
Return ONLY the pattern statement, nothing else.
If these events don't form a meaningful pattern, respond with exactly: NONE

Events:
${contents}`;

    let pattern: string;
    try {
      pattern = await llmChat(prompt);
    } catch {
      continue;
    }

    pattern = pattern.trim();
    if (!pattern || pattern.toUpperCase() === 'NONE') continue;

    let vectors: number[][];
    try {
      vectors = await embedder.embed([pattern]);
    } catch {
      continue;
    }
    if (vectors.length === 0) continue;

    const patternFact: Fact = {
      uuid: randomUUID(),
      content: pattern,
      embedding: vectors[0],
      factType: 'pattern',
      temporalStatus: 'current',
      significance: 5,
      contentKey: defaultContentKey(pattern),
      tag,
      slot: '',
      confidence: 1.0,
      embeddingModel: embedder.modelName(),
      sourceRole: '',
      reinforcedCount: 0,
      recallCount: 0,
    };

    try {
      await store.insertFact(entityUuid, patternFact);
    } catch {
      continue;
    }

    for (const f of group) {
      try {
        await store.updateTemporalStatus(f.uuid, 'historical');
      } catch {
        // best-effort
      }
    }

    patternsCreated++;
  }

  return patternsCreated;
}
